README.md
# Number Names

[![build](https://github.com/calteran/number_names_rs/actions/workflows/build.yml/badge.svg?branch=main)](https://github.com/calteran/number_names_rs/actions/workflows/build.yml)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![codecov](https://codecov.io/gh/calteran/number_names_rs/branch/main/graph/badge.svg?token=WVBHQ5O0MX)](https://codecov.io/gh/calteran/number_names_rs)
[![crates.io](https://img.shields.io/crates/v/number-names)](https://img.shields.io/crates/v/number-names)

Number names is a Rust library to provide formatted string names for cardinal and ordinal numbers.

*At this time, only American English is supported, but there are future plans for i18n.*

## Example usage:

 ```rust
assert_eq!(number_names::cardinal(10), "ten");
assert_eq!(number_names::ordinal(10), "tenth");
 ```

See full documentation at [https://docs.rs/number-names/](https://docs.rs/number-names/).

## Contributing

As this is my first project in Rust, I'm sure there are significant improvements to be made in both the algorithms
and implementation.  I will gladly accept any constructive criticisms, suggestions or pull requests that make
this small project more efficient or accurate.

More specific needs include expanding the library to more languages.  To add a language, create a file in the
/languages folder with the [IETF language tag](https://en.wikipedia.org/wiki/IETF_language_tag) for the language,
in snake_case.  Implement the `cardinal_with_format` and `ordinal_with_format` public functions, returning `String`s
for each `u64` number.  Add a tests module (either separately or in the same file) with tests covering several generic
as well as specific edge-cases for the language.  Add an option to the `number_names::languages::Language` enum with
the IETF tag in CamelCase.  Send me a pull request with the completed changes once all your tests are passing!

## Roadmap

With the release of version 0.2.0 that saw significant refactoring to make the library more extensible for both
formatting and languages, my hope is that the API is now relatively stable.  I hope developers will open issues
or submit enhancements to help strengthen these foundations.  I will continue to maintain this library and intend to
promote the most current version to 1.0 to signify stability no sooner than 6 months after the most recent
breaking change.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0] 2021-10-15
### Changed
 - [**BREAKING API CHANGE**]: Removed the `NumberNames` struct; calls for names are now into simple functions
 - Rebuilt the algorithm for generating cardinal names
 - Moved name generating algorithms into specific `en_us` module to make room for future language additions

### Added
 - `Language` enum to specify the language desired for the given number name
 - `Format` struct, currently only configures the language

## [0.1.2] 2021-10-05
### Fixed
 - Changed license identifier to be picked up correctly by crates.io

## [0.1.1] 2021-10-05
### Added
 - Codecov build script
 - Separate changelog (this file)
### Changed
 - String concatenation style for ordinal numbers

## [0.1.0] 2021-09-24
### Added
- positive cardinal formatting up to `u64::MAX()` in en_US
- positive ordinal formatting up to `u64::MAX()` in en_US


src/tests.rs
#[cfg(test)]
mod tests {
    use crate::*;
    use rstest::rstest;

    #[rstest]
    #[case(0, "zero")]
    #[case(1, "one")]
    #[case(2, "two")]
    #[case(3, "three")]
    #[case(4, "four")]
    #[case(5, "five")]
    #[case(6, "six")]
    #[case(7, "seven")]
    #[case(8, "eight")]
    #[case(9, "nine")]
    #[case(10, "ten")]
    #[case(11, "eleven")]
    #[case(12, "twelve")]
    #[case(13, "thirteen")]
    #[case(14, "fourteen")]
    #[case(15, "fifteen")]
    #[case(16, "sixteen")]
    #[case(17, "seventeen")]
    #[case(18, "eighteen")]
    #[case(19, "nineteen")]
    #[case(20, "twenty")]
    #[case(30, "thirty")]
    #[case(34, "thirty-four")]
    #[case(40, "forty")]
    #[case(50, "fifty")]
    #[case(60, "sixty")]
    #[case(70, "seventy")]
    #[case(80, "eighty")]
    #[case(90, "ninety")]
    #[case(100, "one hundred")]
    #[case(567, "five hundred sixty-seven")]
    #[case(8_910, "eight thousand nine hundred ten")]
    #[case(11_121, "eleven thousand one hundred twenty-one")]
    #[case(314_151, "three hundred fourteen thousand one hundred fifty-one")]
    #[case(
        6_171_819,
        "six million one hundred seventy-one thousand eight hundred nineteen"
    )]
    #[case(
        20_212_223,
        "twenty million two hundred twelve thousand two hundred twenty-three"
    )]
    #[case(
        242_526_272,
        "two hundred forty-two million five hundred twenty-six thousand two hundred seventy-two"
    )]
    #[case(8_293_031_323, "eight billion two hundred ninety-three million thirty-one thousand three hundred twenty-three")]
    #[case(33_435_363_738, "thirty-three billion four hundred thirty-five million three hundred sixty-three thousand seven hundred thirty-eight")]
    #[case(394_041_424_344, "three hundred ninety-four billion forty-one million four hundred twenty-four thousand three hundred forty-four")]
    #[case(4_546_474_849_505, "four trillion five hundred forty-six billion four hundred seventy-four million eight hundred forty-nine thousand five hundred five")]
    #[case(15_253_545_556_575, "fifteen trillion two hundred fifty-three billion five hundred forty-five million five hundred fifty-six thousand five hundred seventy-five")]
    #[case(859_606_162_636_465, "eight hundred fifty-nine trillion six hundred six billion one hundred sixty-two million six hundred thirty-six thousand four hundred sixty-five")]
    #[case(6_667_686_970_717_273, "six quadrillion six hundred sixty-seven trillion six hundred eighty-six billion nine hundred seventy million seven hundred seventeen thousand two hundred seventy-three")]
    #[case(74_757_677_787_980_818, "seventy-four quadrillion seven hundred fifty-seven trillion six hundred seventy-seven billion seven hundred eighty-seven million nine hundred eighty thousand eight hundred eighteen")]
    #[case(283_848_586_878_889_909, "two hundred eighty-three quadrillion eight hundred forty-eight trillion five hundred eighty-six billion eight hundred seventy-eight million eight hundred eighty-nine thousand nine hundred nine")]
    #[case(1_929_394_959_697_989_910, "one quintillion nine hundred twenty-nine quadrillion three hundred ninety-four trillion nine hundred fifty-nine billion six hundred ninety-seven million nine hundred eighty-nine thousand nine hundred ten")]
    #[case(18_446_744_073_709_551_615, "eighteen quintillion four hundred forty-six quadrillion seven hundred forty-four trillion seventy-three billion seven hundred nine million five hundred fifty-one thousand six hundred fifteen")]
    #[case(1_000_000_000, "one billion")]
    fn cardinal_name(#[case] input: u64, #[case] expected: String) {
        assert_eq!(expected, cardinal(input))
    }

    #[test]
    fn ordinal_name() {
        let values = values();

        for value in values {
            assert_eq!(
                value.2.to_string(),
                ordinal(value.0),
                "Failed on {}",
                value.0
            );
        }
    }

    fn values() -> [(u64, &'static str, &'static str); 49] {
        [
            (0,
             "zero",
             "zeroth"),
            (1,
             "one",
             "first"),
            (2,
             "two",
             "second"),
            (3,
             "three",
             "third"),
            (4,
             "four",
             "fourth"),
            (5,
             "five",
             "fifth"),
            (6,
             "six",
             "sixth"),
            (7,
             "seven",
             "seventh"),
            (8,
             "eight",
             "eighth"),
            (9,
             "nine",
             "ninth"),
            (10,
             "ten",
             "tenth"),
            (11,
             "eleven",
             "eleventh"),
            (12,
             "twelve",
             "twelfth"),
            (13,
             "thirteen",
             "thirteenth"),
            (14,
             "fourteen",
             "fourteenth"),
            (15,
             "fifteen",
             "fifteenth"),
            (16,
             "sixteen",
             "sixteenth"),
            (17,
             "seventeen",
             "seventeenth"),
            (18,
             "eighteen",
             "eighteenth"),
            (19,
             "nineteen",
             "nineteenth"),
            (20,
             "twenty",
             "twentieth"),
            (30,
             "thirty",
             "thirtieth"),
            (34,
             "thirty-four",
             "thirty-fourth"),
            (40,
             "forty",
             "fortieth"),
            (50,
             "fifty",
             "fiftieth"),
            (60,
             "sixty",
             "sixtieth"),
            (70,
             "seventy",
             "seventieth"),
            (80,
             "eighty",
             "eightieth"),
            (90,
             "ninety",
             "ninetieth"),
            (100,
             "one hundred",
             "one hundredth"),
            (567,
             "five hundred sixty-seven",
             "five hundred sixty-seventh"),
            (8_910,
             "eight thousand nine hundred ten",
             "eight thousand nine hundred tenth"),
            (11_121,
             "eleven thousand one hundred twenty-one",
             "eleven thousand one hundred twenty-first"),
            (314_151,
             "three hundred fourteen thousand one hundred fifty-one",
             "three hundred fourteen thousand one hundred fifty-first"),
            (6_171_819,
             "six million one hundred seventy-one thousand eight hundred nineteen",
             "six million one hundred seventy-one thousand eight hundred nineteenth"),
            (20_212_223,
             "twenty million two hundred twelve thousand two hundred twenty-three",
             "twenty million two hundred twelve thousand two hundred twenty-third"),
            (242_526_272,
             "two hundred forty-two million five hundred twenty-six thousand two hundred seventy-two",
             "two hundred forty-two million five hundred twenty-six thousand two hundred seventy-second"),
            (8_293_031_323,
             "eight billion two hundred ninety-three million thirty-one thousand three hundred twenty-three",
             "eight billion two hundred ninety-three million thirty-one thousand three hundred twenty-third"),
            (33_435_363_738,
             "thirty-three billion four hundred thirty-five million three hundred sixty-three thousand seven hundred thirty-eight",
             "thirty-three billion four hundred thirty-five million three hundred sixty-three thousand seven hundred thirty-eighth"),
            (394_041_424_344,
             "three hundred ninety-four billion forty-one million four hundred twenty-four thousand three hundred forty-four",
             "three hundred ninety-four billion forty-one million four hundred twenty-four thousand three hundred forty-fourth"),
            (4_546_474_849_505,
             "four trillion five hundred forty-six billion four hundred seventy-four million eight hundred forty-nine thousand five hundred five",
             "four trillion five hundred forty-six billion four hundred seventy-four million eight hundred forty-nine thousand five hundred fifth"),
            (15_253_545_556_575,
             "fifteen trillion two hundred fifty-three billion five hundred forty-five million five hundred fifty-six thousand five hundred seventy-five",
             "fifteen trillion two hundred fifty-three billion five hundred forty-five million five hundred fifty-six thousand five hundred seventy-fifth"),
            (859_606_162_636_465,
             "eight hundred fifty-nine trillion six hundred six billion one hundred sixty-two million six hundred thirty-six thousand four hundred sixty-five",
             "eight hundred fifty-nine trillion six hundred six billion one hundred sixty-two million six hundred thirty-six thousand four hundred sixty-fifth"),
            (6_667_686_970_717_273,
             "six quadrillion six hundred sixty-seven trillion six hundred eighty-six billion nine hundred seventy million seven hundred seventeen thousand two hundred seventy-three",
             "six quadrillion six hundred sixty-seven trillion six hundred eighty-six billion nine hundred seventy million seven hundred seventeen thousand two hundred seventy-third"),
            (74_757_677_787_980_818,
             "seventy-four quadrillion seven hundred fifty-seven trillion six hundred seventy-seven billion seven hundred eighty-seven million nine hundred eighty thousand eight hundred eighteen",
             "seventy-four quadrillion seven hundred fifty-seven trillion six hundred seventy-seven billion seven hundred eighty-seven million nine hundred eighty thousand eight hundred eighteenth"),
            (283_848_586_878_889_909,
             "two hundred eighty-three quadrillion eight hundred forty-eight trillion five hundred eighty-six billion eight hundred seventy-eight million eight hundred eighty-nine thousand nine hundred nine",
             "two hundred eighty-three quadrillion eight hundred forty-eight trillion five hundred eighty-six billion eight hundred seventy-eight million eight hundred eighty-nine thousand nine hundred ninth"),
            (1_929_394_959_697_989_910,
             "one quintillion nine hundred twenty-nine quadrillion three hundred ninety-four trillion nine hundred fifty-nine billion six hundred ninety-seven million nine hundred eighty-nine thousand nine hundred ten",
             "one quintillion nine hundred twenty-nine quadrillion three hundred ninety-four trillion nine hundred fifty-nine billion six hundred ninety-seven million nine hundred eighty-nine thousand nine hundred tenth"),
            (18_446_744_073_709_551_615,
             "eighteen quintillion four hundred forty-six quadrillion seven hundred forty-four trillion seventy-three billion seven hundred nine million five hundred fifty-one thousand six hundred fifteen",
             "eighteen quintillion four hundred forty-six quadrillion seven hundred forty-four trillion seventy-three billion seven hundred nine million five hundred fifty-one thousand six hundred fifteenth"),
            (1_000_000_000,
             "one billion",
             "one billionth")
        ]
    }
}


src/lib.rs
#![warn(missing_docs)]
#![warn(rustdoc::missing_doc_code_examples)]
#![doc = include_str!("../README.md")]

pub mod format;
pub mod languages;

use crate::format::Format;
use crate::languages::Language::*;

/// Return cardinal number name for a given number with the default format
///
/// Example:
/// ```rust
/// assert_eq!(number_names::cardinal(10), "ten".to_string())
/// ```
pub fn cardinal(num: u64) -> String {
    cardinal_with_fmt(num, Format::default())
}

/// Return cardinal number name for a given number with a given format
///
/// Example:
/// ```rust
/// use number_names::{cardinal_with_fmt, Format};
///
/// assert_eq(cardinal_with_fmt(20, Format::default()), "twenty".to_string());
/// ```
pub fn cardinal_with_fmt(num: u64, fmt: Format) -> String {
    match fmt.language {
        EnUs => crate::languages::en_us::cardinal_with_format(num, fmt),
        _ => crate::languages::en_us::cardinal_with_format(num, fmt),
    }
}

/// Return ordinal number name for a given number with the default format
///
/// Example:
/// ```rust
/// assert_eq!(number_names::ordinal(10), "tenth".to_string())
/// ```
pub fn ordinal(num: u64) -> String {
    ordinal_with_fmt(num, Format::default())
}

/// Return ordinal number name for a given number with a given format
///
/// Example:
/// ```rust
/// use number_names::{ordinal_with_fmt, Format};
///
/// assert_eq(ordinal_with_fmt(20, Format::default()), "twentieth".to_string());
/// ```
pub fn ordinal_with_fmt(num: u64, fmt: Format) -> String {
    match fmt.language {
        EnUs => crate::languages::en_us::ordinal_with_format(num, fmt),
        _ => crate::languages::en_us::ordinal_with_format(num, fmt),
    }
}

mod tests;


src/format.rs
//! Defines configuration struct Format to control number name formatting.
use crate::languages::Language;
use crate::languages::Language::EnUs;

///
/// Configuration struct to control formatting of the number names.
/// Currently used only to specify the language.
///
/// Examples:
///
/// ```rust
/// // default format
/// let fmt = number_names::Format::default();
///
/// // specify a language
/// let fmt = number_names::Format(number_names::languages::Language::EnUs);
/// ```
pub struct Format {
    /// Language the number name should be formatted in
    pub language: Language,
}

impl Format {
    /// Return a Format with the default values
    /// (currently only language set to American English)
    ///
    /// Example:
    /// ```rust
    /// let fmt = number_names::Format::default();
    /// assert_eq!(fmt, number_names::Format{number_names::languages::Language::EnUs});
    /// ```
    pub fn default() -> Format {
        Format { language: EnUs }
    }
}


src/languages/en_us.rs
//! Generate number names in American English
use crate::format::Format;
use std::cmp;

/// Return cardinal number name for a given number with a given format in American English
///
/// Example:
/// ```rust
/// use number_names::{languages::en_us::cardinal_with_fmt, Format};
///
/// assert_eq(cardinal_with_fmt(20, Format::default()), "twenty".to_string());
/// ```
pub fn cardinal_with_format(num: u64, fmt: Format) -> String {
    encode(num, fmt)
}

/// Return cardinal number name for a given number with a given format in American English
///
/// Example:
/// ```rust
/// use number_names::{languages::en_us::ordinal_with_fmt, Format};
///
/// assert_eq(ordinal_with_fmt(20, Format::default()), "twentieth".to_string());
/// ```
pub fn ordinal_with_format(num: u64, fmt: Format) -> String {
    // Lets start with the cardinal version and then replace the last word
    let cardinal = cardinal_with_format(num, fmt);

    // find the character (this is all in ASCII!) before the last word
    // it might be a space or a dash (eg. "thirty-three")
    let last_break = match (cardinal.rfind(' '), cardinal.rfind('-')) {
        (Some(space), Some(dash)) => cmp::max(space, dash) + 1,
        (Some(space), None) => space + 1,
        (None, Some(dash)) => dash + 1,
        (None, None) => 0,
    };

    // Most numbers just add "-th" to the end of the cardinal name, but others are special
    match NOT_TH
        .iter()
        .position(|(word, _)| word.eq(&&cardinal[last_break..]))
    {
        Some(index) => String::from(&cardinal[..last_break]) + NOT_TH[index].1,
        None => cardinal + "th",
    }
}

fn encode(num: u64, fmt: Format) -> String {
    let mut num = num;
    if num > 0 {
        std::iter::from_fn(|| {
            if num > 0 {
                let v = num % 1000;
                num /= 1000;
                Some(v)
            } else {
                None
            }
        })
        .map(|num| format_num(num, &fmt))
        .zip(ORDERS.iter())
        .filter(|(num, _)| num != "zero")
        .map(|(num, order)| num + order)
        .collect::<Vec<String>>()
        .into_iter()
        .rev()
        .collect::<Vec<String>>()
        .join(" ")
    } else {
        "zero".to_string()
    }
}

fn format_num(num: u64, fmt: &Format) -> String {
    match num {
        0..=19 => ONES[num as usize].to_string(),
        20..=99 => {
            let upper = (num / 10) as usize;
            match num % 10 {
                0 => TENS[upper].to_string(),
                lower => format!("{}-{}", TENS[upper], ONES[lower as usize]),
            }
        }
        _ => {
            let mut v = vec![
                format_num(num / 100, fmt),
                "hundred".to_string(),
                format_num(num % 100, fmt),
            ];
            v.retain(|s| s != "zero");
            v.join(" ")
        }
    }
}

const ONES: [&str; 20] = [
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
    "ten",
    "eleven",
    "twelve",
    "thirteen",
    "fourteen",
    "fifteen",
    "sixteen",
    "seventeen",
    "eighteen",
    "nineteen",
];

const TENS: [&str; 10] = [
    "zero", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety",
];

const ORDERS: [&str; 7] = [
    "",
    " thousand",
    " million",
    " billion",
    " trillion",
    " quadrillion",
    " quintillion",
];

const NOT_TH: [(&str, &str); 15] = [
    ("one", "first"),
    ("two", "second"),
    ("three", "third"),
    ("five", "fifth"),
    ("eight", "eighth"),
    ("nine", "ninth"),
    ("twelve", "twelfth"),
    ("twenty", "twentieth"),
    ("thirty", "thirtieth"),
    ("forty", "fortieth"),
    ("fifty", "fiftieth"),
    ("sixty", "sixtieth"),
    ("seventy", "seventieth"),
    ("eighty", "eightieth"),
    ("ninety", "ninetieth"),
];
